From c43498e8cf6fb66ee025a0e7094cede4ee48b27f Mon Sep 17 00:00:00 2001
From: Kjell Ahlstedt <kjellahlstedt@gmail.com>
Date: Sun, 5 Jul 2026 16:23:42 +0200
Subject: [PATCH 1/2] Gio: Emblem and DBus::ActionGroup: Don't derive
 gtkmm__Gxxx types

GEmblem and GDBusActionGroup are declared G_DECLARE_FINAL_TYPE.
Add _DO_NOT_DERIVE_GTYPE, _ABI_AS_WITH_DERIVED_GTYPE and _STRUCT_NOT_HIDDEN
to the corresponding glibmm classes.
---
 gio/src/dbusactiongroup.hg | 14 ++++++++++++++
 gio/src/emblem.hg          |  4 ++++
 2 files changed, 18 insertions(+)

diff --git a/gio/src/dbusactiongroup.hg b/gio/src/dbusactiongroup.hg
index fbc00f86..fedba12b 100644
--- a/gio/src/dbusactiongroup.hg
+++ b/gio/src/dbusactiongroup.hg
@@ -21,6 +21,17 @@
 _DEFS(giomm,gio)
 _PINCLUDE(glibmm/private/object_p.h)
 
+#ifndef DOXYGEN_SHOULD_SKIP_THIS
+// Necessary if compiled against glib < 2.89.1, where GDBusActionGroup is not
+// a final type, and GDBusActionGroupClass is not defined by G_DECLARE_FINAL_TYPE().
+// (It was declared a final type somewhere between 2.89.1 and 2.89.2.
+// The compiler will probably complain if you compile against a glib 2.89.1
+// where GDBusActionGroup is not a final type.)
+#if !GLIB_CHECK_VERSION(2,89,1)
+using GDBusActionGroupClass = struct _GDBusActionGroupClass;
+#endif
+#endif /* DOXYGEN_SHOULD_SKIP_THIS */
+
 namespace Gio
 {
 
@@ -43,6 +54,9 @@ class GIOMM_API ActionGroup
   _CLASS_GOBJECT(ActionGroup, GDBusActionGroup, G_DBUS_ACTION_GROUP, Glib::Object, GObject, , , GIOMM_API)
   _IMPLEMENTS_INTERFACE(Gio::ActionGroup)
   _IMPLEMENTS_INTERFACE(Gio::RemoteActionGroup)
+  _DO_NOT_DERIVE_GTYPE dnl// GDBusActionGroup is a final type since glib 2.89.2
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
+  _STRUCT_NOT_HIDDEN
 
 protected:
   _CTOR_DEFAULT
diff --git a/gio/src/emblem.hg b/gio/src/emblem.hg
index e8e8cfb4..2e6412f8 100644
--- a/gio/src/emblem.hg
+++ b/gio/src/emblem.hg
@@ -16,6 +16,7 @@
 
 #include <glibmm/object.h>
 #include <giomm/icon.h>
+#include <gio/gio.h>
 
 _DEFS(giomm,gio)
 _PINCLUDE(glibmm/private/object_p.h)
@@ -40,6 +41,9 @@ class GIOMM_API Emblem
 {
   _CLASS_GOBJECT(Emblem, GEmblem, G_EMBLEM, Glib::Object, GObject, , , GIOMM_API)
   _IMPLEMENTS_INTERFACE(Icon)
+  _DO_NOT_DERIVE_GTYPE dnl// GEmblem is a final type since glib 2.89.2
+  _ABI_AS_WITH_DERIVED_GTYPE dnl// Remove when we can break ABI
+  _STRUCT_NOT_HIDDEN
 
 public:
   _WRAP_ENUM(Origin, GEmblemOrigin, NO_GTYPE, decl_prefix GIOMM_API)
-- 
2.54.0

